core/src/main/scala/cats/data/Chain.scala

```diff
@@ -189,10 +189,33 @@ sealed abstract class Chain[+A] extends ChainCompat[A] {
    * Folds over the elements from left to right using the supplied initial value and function.
    */
   final def foldLeft[B](z: B)(f: (B, A) => B): B = {
-    var result = z
-    val iter = iterator
-    while (iter.hasNext) { result = f(result, iter.next()) }
-    result
+    @annotation.tailrec
+    def loop(h: Chain.NonEmpty[A], tail: List[Chain.NonEmpty[A]], acc: B): B =
```

Comment:
by passing the destructured list, we avoid an allocation in the `Append` case (which is common since Append is the way we combine two chains).